### ISSUE TITLE
I try softmoe in my vision classification task, but in fact the effect cannot even be compared to baseline

### ISSUE THREAD

**Issue by daixiangzi** [google-research/vmoe#159]
I use code:https://github.com/lucidrains/soft-moe-pytorch

**Comment by daixiangzi** [comment]
1000w train data in vit-b-32 on last 6 layers with 64-experts softmoe

**Comment by daixiangzi** [comment]
can you give me some train softmoe advise

**Comment by jpuigcerver** [comment]
We are unable to provide help for a different implementation, since we are not familiar with their internal details.

One thing to keep in mind is that MoE models in general, and Soft MoEs in particular, have a massive number of parameters, and are quite data hungry. So, if you are training it from scratch, make sure that you have a pre-training dataset large enough.

**Comment by daixiangzi** [comment]
> We are unable to provide help for a different implementation, since we are not familiar with their internal details.
> 
> One thing to keep in mind is that MoE models in general, and Soft MoEs in particular, have a massive number of parameters, and are quite data hungry. So, if you are training it from scratch, make sure that you have a pre-training dataset large enough.

ok, in addition to the particularity of the data, are there any special aspects during your training, such as the learning rate of the moe layer is different from other layers, or the initialization is different?

**Comment by jpuigcerver** [comment]
Not really. We use the same learning rate for all the model. We use weight decay, but we apply the same factors to MoE and non-MoE layers, the only difference is in the head vs. rest of the network.

You can take a look at the config file of the Soft-MoE pre-training on JFT in [vmoe/projects/soft_moe/configs/pretrain_jft4b.py](https://github.com/google-research/vmoe/blob/main/vmoe/projects/soft_moe/configs/pretrain_jft4b.py). The hparams are the same whether you use Soft MoE or other types of MoEs (e.g. Experts Choice, or vanilla Top-K MoE).

We tested a very similar config on LAION-400M (essentially same hparams), and works very well there as well. We'll upload it soon, once the ICLR revisions are completed.

**Comment by daixiangzi** [comment]
> pretrain_jft4b.py

That’s very strange. My experiment here is not a visual classification task, but a visual comparison learning task similar to clip single branch. The final output of the network is not the classification head, but the representation embedding. Mainly I saw you The pretrain_jft4b configuration is similar to mine. I use adamw+PolynomialLRWarmup+0.001+weight_decay_ 0.05

**Comment by daixiangzi** [comment]
Actually, I'm using **pre_ Logits** instead of **head**, I see in your configuration file the head and weight_ decay is 3, so should I also be in pre_logits weight_decay should also be 3?

**Comment by jpuigcerver** [comment]
Mmm. I'm not familiar with the codebase that you use, but I guess that you are using an extra linear projection before the head. We haven't used that in a while.

In any case, when we did use a "pre_logits projection", we still used the same weight decay as in the rest of the model. We only change the weight decay of the very last linear projection (i.e. the head).

Also, this is mostly done for transfer purposes. If you only care about the pretraining performance, probably using the same weight decay for everything works fine.

**Comment by daixiangzi** [comment]
> Mmm. I'm not familiar with the codebase that you use, but I guess that you are using an extra linear projection before the head. We haven't used that in a while.
> 
> In any case, when we did use a "pre_logits projection", we still used the same weight decay as in the rest of the model. We only change the weight decay of the very last linear projection (i.e. the head).
> 
> Also, this is mostly done for transfer purposes. If you only care about the pretraining performance, probably using the same weight decay for everything works fine.

Another point is that I see that you did not add a cls token by default. At the end of the network, the average value of all tokens was taken by default. This seems to be a little different from the original vit. Are you doing this to obtain better performance? ?

**Comment by daixiangzi** [comment]
The default classifier adopts the gap method:https://github.com/google-research/vmoe/blob/dfb9ee01ce6dfc5a8228b406e768ee325dd18fcd/vmoe/nn/vit_moe.py#L361
